### PR TITLE
[SPARK-35294][SQL] Add tree traversal pruning in rules with dedicated files under optimizer

### PR DESCRIPTION
**Pull request by sigmod**
### What changes were proposed in this pull request?

Added the following TreePattern enums:
- CREATE_NAMED_STRUCT
- EXTRACT_VALUE
- JSON_TO_STRUCT
- OUTER_REFERENCE
- AGGREGATE
- LOCAL_RELATION
- EXCEPT
- LIMIT
- WINDOW

Used them in the following rules:
- DecorrelateInnerQuery
- LimitPushDownThroughWindow
- OptimizeCsvJsonExprs
- PropagateEmptyRelation
- PullOutGroupingExpressions
- PushLeftSemiLeftAntiThroughJoin
- ReplaceExceptWithFilter
- RewriteDistinctAggregates
- SimplifyConditionalsInPredicate
- UnwrapCastInBinaryComparison

### Why are the changes needed?

Reduce the number of tree traversals and hence improve the query compilation latency.

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by sigmod** [comment]
@hvanhovell @gengliangwang @dbaliafroozeh @maryannxue this PR is ready for review. Let me know if you have any questions.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LocalRelation.scala [hunk 1]
```diff
@@ -93,4 +94,12 @@ case class LocalRelation(
   }
 
   override def maxRows: Option[Long] = Some(data.length.toLong)
+
+  override val nodePatterns: Seq[TreePattern] = {
+    if (this.data.isEmpty) {
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LocalRelation.scala:99]: shall we just make it `LOCAL_RELATION`?
We can prune `LOCAL_RELATION` in `ConvertToLocalRelation`.
We got general patterns like `BINARY_COMPARISON` and `BINARY_ARITHMETIC`. This one is very specific and it looks odd.

**Review by gengliangwang — APPROVED** [review]
LGTM except one comment

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LocalRelation.scala (hunk 1 above)
Comment (sigmod) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LocalRelation.scala]: Done. 
Thanks, Gengliang! Make sense!

**Comment by gengliangwang** [comment]
Thanks, merging to master

**Pull request closed without merging**